Add jsonld canonization

krobi64 opened the pull request:
[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

### Release

Semantic release is enabled for this repository. Make sure you follow the right commit message convention. 

We're using semantic-release's default — [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).

## Description of Changes

Resolves [frost-api: 243](https://github.com/poetapp/frost-api/issues/243)

* **refactor**: Implemented jsonld library to canonize documents to generate claim ids.

Note, there is a performance change that will need to occur: caching the context references to prevent retrieval on each call.

ericelliott commented:
Does this one PR resolve the whole context of frost-api: 243?

ericelliott commented on src/Claim.test.ts:
```diff
@@ -41,31 +54,117 @@ const PrivateKeyEAP = 'KxuZJmgVAipi9hfYXHTyGYmmhkbG7fBzmkyVnj6t9j9rDR1nN1vN'
+  }
+
+  {
+    try {
```
Why are we wrapping `try/catch` around the entire assert block? Does the `createClaim(Key.privateKey, ClaimType.Work, TheRaven.attributes).catch(err => err)` trick not work here?

ericelliott commented on src/Claim.test.ts:
```diff
@@ -41,31 +54,117 @@ const PrivateKeyEAP = 'KxuZJmgVAipi9hfYXHTyGYmmhkbG7fBzmkyVnj6t9j9rDR1nN1vN'
+  }
+
+  {
+    try {
```
Ditto

ericelliott commented on src/Claim.ts:
```diff
@@ -57,20 +68,24 @@ export function createClaim(privateKey: string, type: ClaimType, attributes: Cla
-  const id = getClaimId(claim)
-  const signature = getClaimSignature(
-    {
+  const id = await getClaimId(claim)
```
Isn't the id a multihash of the claim?

If it is, don't we just need to sign the id, and not the whole claim + id?

krobi64 replied:
> Does this one PR resolve the whole context of frost-api: 243?

>Canonicalize claims for hashing in Frost API #243

I realized as I built out my lengthy todo list, that most of it was outside the scope of the ticket as written.

We need to capture everything else, but this PR does do what the ticket is asking. (Once I modify as requested.

krobi64 replied:
It is just signing the id. However before that happens there is a fair amount of validation that occurs and requires the claim itself. Refer to line 46 of this file. Validations include:

* Is claim.publicKey present?
* Does the provided private key prove to verify the against the provided public key?
* Is clam.id present?
* Does a generated claim id match claim.id?

I didn't want to break this out without a lot more thought.

lautarodragan replied:
> Isn't the id a multihash of the claim?

Now that you mention it, no, it isn't, but it shouldn't be too complicated to make it so. I've used the multihash lib and it's straight forward. 

Do we want to make the id a multihash (a breaking change) before mainnet beta? @ericelliott (We'd still address it in a separate PR)

ericelliott replied:
Let's consider this carefully. Is it possible two have two identical valid claims?

Put another way, would it ever make sense for a single entity to say the same thing about the same subject at a time such that the two identical claims have some kind of cumulative effect?

lautarodragan commented on typings/jsonld.d.ts:
```diff
@@ -0,0 +1 @@
+declare module 'jsonld'
```
Can you provide typings for the parts of jsonld being used in this PR?

krobi64 replied:
Can you provide some guidance? Point me to an example?

Here's the context:

I’m using two functions from jsonld: `expand` and `canonize` (so far). Eventually the plan is to also add `sign` and `verify`, what do I need to define in the type file for jsonld (since it isn’t typescript)?

`canonize` returns a string form of the document (edited)
`expand` returns an object in which all of the keys have been expanded to the url definition of each attribute of the document, per the provided context.

e.g.
  ```{
    'http://schema.org/CreativeWork': [
      {
        'http://schema.org/author': [
          {
            '@value': 'Edgar Allan Poe',
          },
        ],
        'http://schema.org/dateCreated': [
          {
            '@value': '',
          },
        ],
        'http://schema.org/datePublished': [
          {
            '@value': '1845-01-29T03:00:00.000Z',
          },
        ],
        'http://schema.org/name': [
          {
            '@value': 'The Raven',
          },
        ],
        'http://schema.org/keyword': [
          {
            '@value': 'poem',
          },
        ],
        'http://schema.org/text': [
          {
            '@value': 'Once upon a midnight dreary...',
          },
        ],
      },
    ],
    'http://purl.org/dcterms/created': [
      {
        '@value': 'Mon Nov 13 2017 15:00:00 GMT+0000 (UTC)',
      },
    ],
    'http://schema.org/Text': [
      {
        '@value': '02badf4650ba545608242c2d303d587cf4f778ae3cf2b3ef99fbda37555a400fd2',
      },
    ],
    'http://schema.org/additionalType': [
      {
        '@value': 'Work',
      },
    ],
  },
]
```

ericelliott replied:
I think @jwicks31 and @kennylavender have experience adding typings for libraries not written in TypeScript (I believe that's how we added support for the feature toggle libraries). Maybe one of them can help?

krobi64 commented on src/Claim.ts:
```diff
@@ -1,42 +1,64 @@
-    .update(buffer)
-    .digest()
-    .toString('hex')
+loadContext()
```
Removed this for now. I need to track down how `expand` is loading context. I think it has a network-related side effect.

lautarodragan commented on src/Claim.ts:
```diff
@@ -1,42 +1,62 @@
 
-export function getClaimSignature(claim: Claim, privateKey: string): string {
+export const getClaimId = async (claim: Claim): Promise<string> => {
+  try {
```
I think this try/catch is not doing anything...

krobi64 replied:
ack

lautarodragan commented on src/Claim.ts:
```diff
@@ -48,29 +68,33 @@ export function isValidSignature(claim: Claim): boolean {
-  const signature = getClaimSignature(
-    {
+  const id = await getClaimId(claim)
+  try {
```
Same with this one

krobi64 replied:
removed

lautarodragan commented on src/Claim.ts:
```diff
@@ -1,42 +1,62 @@
 import * as crypto from 'crypto'
-
+import { canonize, expand } from 'jsonld'
 import { IllegalArgumentException } from './Exceptions'
```
Please leave an empty line between 3rd party imports and imports of the same project

krobi64 replied:
ack

krobi64 commented:
@ericelliott While debugging some failing tests in the poet-node application, Lautaro and I found some tests failing specifically due to performance. I decided to take Friday and look at it more deeply.

Here are the numbers just running a simple canonization vs the one used in jsonld (and its signature suites):

$ time node json-perftest.js
real	0m1.649s
user	0m1.846s
sys	0m0.124s

$ time node jsonld-cached-perftest.js
real	0m45.841s
user	0m44.547s
sys	0m1.658s

I believe there can be significant performance using a cached version of the context we have to supply with any jsonld document. However, I haven't gotten the example working for our use case. I have a ticket open for the library's authors to look at (https://github.com/digitalbazaar/jsonld.js/issues/261)

lautarodragan commented:
@krobi64 do we want to mark this PR as blocked for now, or should we go ahead with it and optimize later? 

If we know for sure we can optimize later, I'd rather leave that as a separate PR.

krobi64 commented:
I believe that we can definitely optimize. I would prefer to optimize later so that we can move forward.

krobi64 commented:
@ericelliott no more review is required. I moved the location of the `context` data into Interfaces since it is mostly tied there. And I added some developer notes to clarify its purpose.

lautarodragan commented on typings/jsonld.d.ts:
```diff
@@ -0,0 +1,4 @@
+declare module 'jsonld' {
+  export function canonize(doc: any): string
```
Doesn't this return a `Promise<string>`?

lautarodragan commented on typings/jsonld.d.ts:
```diff
@@ -0,0 +1,4 @@
+declare module 'jsonld' {
+  export function canonize(doc: any): string
+  export function expand(doc: any): any
```
Just to make sure I'm not missing anything — is `expand` not being used anywhere? (Nothing wrong with that, just asking)

lautarodragan commented on src/Interfaces.ts:
```diff
@@ -17,10 +16,30 @@ export function isClaim(object: any): object is Claim {
+// WARNING: This MUST account for ALL of the attributes in a Claim, except for id, @context, and signature.
+// Otherwise those attributes without context will be left out of the canonized/signed claim.
+// Refer to https://www.w3.org/2018/jsonld-cg-reports/json-ld/#the-context
+export function getClaimContext(): any {
```
We could declare this as a `const` instead of a function and mark all fields as `readonly`.

```ts
interface ClaimContext {
    readonly '@context': {
      readonly [key: string]: string
    }
}

const ClaimContext: ClaimContext = {
    '@context': {
        attributes: 'http://schema.org/CreativeWork',
        // etc
    }
}
```

Attempting to mutate ClaimContext in any way will result in an error:

```
index.ts:14:1 - error TS2542: Index signature in type '{ readonly [key: string]: string; }' only permits reading.
```

wzalazar commented on .travis.yml:
```diff
@@ -1,5 +1,10 @@
 language: node_js
 before_install:
+  - sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
```
I have a question, Do we need to update the requirements in our documentation?

krobi64 replied:
@warrenv Good question. I can run this in our local-dev without changing anything, so I didn't think so. It appeared to be a Travis issue you helped me resolve

wzalazar replied:
maybe can be a `know issue`, but I don't know.

kennylavender replied:
I am running Ubuntu as my os and did not need to run this, and I am pretty sure I never have before. Travis is probably running a stripped out version of Ubuntu to reduce server load and they have you add back things.

ericelliott commented on src/Interfaces.test.ts:
```diff
@@ -8,14 +8,14 @@ const TheRaven: Work = {
     '3045022100e020a7ffeffa5d40ffde618c6c861678e38de69fd377028ec57ad93893883b3702201f085284a9064bab7e1cd39349e65d136d8f67e4b6b897c3e7db6b400ed91034',
   type: ClaimType.Work,
-  dateCreated: new Date('2017-11-13T15:00:00.000Z'),
+  created: '2017-11-13T15:00:00.000Z',
```
Can we clarify in our API/docs that we're expecting a string here, not a JS date object?

krobi64 replied:
None of the apis take this as input. We generate this. It IS documented here: https://github.com/poetapp/random/blob/master/Claims.md
Note that I haven't created a PR to replace `created` with `issuanceDate`

ericelliott commented on src/Interfaces.ts:
```diff
@@ -29,17 +45,17 @@ export enum ClaimType {
 export interface WorkAttributes extends ClaimAttributes {
   readonly name: string
   readonly datePublished: string
   readonly dateCreated: string
```
These date types look correct. Why did our test code try to use a JS date object?

kennylavender commented on package.json:
```diff
@@ -8,13 +8,13 @@
     "lint": "tslint -p ./tsconfig.json",
     "lint:fix": "tslint -p ./tsconfig.json --fix",
-    "test:unit": "ts-node --files ./tests/unit/index.ts",
+    "test:unit": "TZ=UTC ts-node --files ./tests/unit/index.ts",
```
What is the reason we need to force a timezone? I find it weird that we need to force a timezone, I ran the tests locally with and without the timezone and it seems to work.

krobi64 replied:
Actually now that `created` is an iso string, it's probably not required. Before the local tests for me came back with PDT and on Travis as UTC. I'll verify

kennylavender replied:
Cool, sounds good, if you are able to remove it, don't forget to remove it from "prepublish" as well

kennylavender commented on src/Interfaces.ts:
```diff
@@ -1,24 +1,40 @@
+    attributes: 'http://schema.org/CreativeWork',
+    author: 'http://schema.org/author',
+    text: 'http://schema.org/text',
+    created: 'http://purl.org/dcterms/created',
```
This is giving a 503 error, it may be correct though, just pointing it out.

krobi64 replied:
ahhh you are correct, it should be http://purl.org/dc/terms/created
 per http://dublincore.org/documents/dcmi-terms/#terms-created

krobi64 commented on src/Interfaces.ts:
```diff
@@ -1,24 +1,40 @@
+    dateCreated: 'http://schema.org/dateCreated',
+    datePublished: 'http://schema.org/datePublished',
+    name: 'http://schema.org/name',
+    tags: 'http://schema.org/keyword',
```
Good catch
